File: frontend/src/lib/ipfs.ts
/**
 * IPFS Upload Service
 * 
 * Uses Pinata's public pinning API (free tier) for decentralized file storage.
 * Files are uploaded as encrypted blobs — the gateway just serves raw bytes.
 */

const PINATA_GATEWAY = process.env.NEXT_PUBLIC_PINATA_GATEWAY || 'https://gateway.pinata.cloud/ipfs';
const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

interface BackendIpfsResponse {
    success: boolean;
    cid: string;
    url: string;
    size: number;
}

/**
 * Upload an encrypted blob to IPFS via the local API Backend proxy.
 * Ensures secret API keys are never exposed to the frontend browser.
 */
export async function uploadToIPFS(
    blob: Blob,
    fileName: string,
    metadata?: Record<string, string>
): Promise<{ cid: string; url: string; size: number }> {
    const formData = new FormData();
    formData.append('file', blob, `${fileName}.encrypted`);

    // Pinata metadata (optional for the proxy, but backend accepts it)
    const pinataMetadata = JSON.stringify({
        name: `medilock-${fileName}`,
        keyvalues: {
            app: 'medilock',
            encrypted: 'true',
            ...metadata,
        },
    });
    formData.append('pinataMetadata', pinataMetadata);

    const pinataOptions = JSON.stringify({
        cidVersion: 1,
    });
    formData.append('pinataOptions', pinataOptions);

    // Call the local backend securely
    const response = await fetch(`${API_URL}/api/ipfs/upload`, {
        method: 'POST',
        body: formData,
    });

    if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`Backend IPFS proxy failed: ${response.status} — ${errorText}`);
    }

    const result: BackendIpfsResponse = await response.json();

    if (!result.success || !result.cid) {
        throw new Error(`Backend IPFS proxy invalid response`);
    }

    return {
        cid: result.cid,
        url: `${PINATA_GATEWAY}/${result.cid}`, // Construct public URL using gateway
        size: result.size,
    };
}

/**
 * Fetch encrypted data from IPFS using a CID.
 * Returns the raw text (base64-encoded ciphertext).
 */
export async function fetchFromIPFS(cid: string): Promise<string> {
    const url = `${PINATA_GATEWAY}/${cid}`;
    const response = await fetch(url);

    if (!response.ok) {
        throw new Error(`IPFS fetch failed: ${response.status}`);
    }

    return response.text();
}

/**
 * Get the public IPFS URL for a CID.
 */
export function getIPFSUrl(cid: string): string {
    return `${PINATA_GATEWAY}/${cid}`;
}

/**
 * Check if IPFS upload is supported.
 * Since credentials are now securely managed on the backend, this is assumed true
 * assuming the backend is running.
 */
export function isIPFSConfigured(): boolean {
    return true;
}

File: backend/routes/ipfs.py
import os
from fastapi import APIRouter, HTTPException, UploadFile, File, Form
import httpx
from pydantic import BaseModel
import json

router = APIRouter()

PINATA_API_KEY = os.getenv("PINATA_API_KEY")
PINATA_SECRET_KEY = os.getenv("PINATA_SECRET_KEY")
PINATA_GATEWAY = os.getenv("PINATA_GATEWAY", "https://gateway.pinata.cloud/ipfs")

@router.post("/ipfs/upload")
async def upload_to_ipfs(
    file: UploadFile = File(...),
    pinataMetadata: str = Form(None),
    pinataOptions: str = Form(None)
):
    """
    Securely proxies the file upload to Pinata using backend credentials.
    Accepts multipart/form-data.
    """
    if not PINATA_API_KEY or not PINATA_SECRET_KEY:
        raise HTTPException(status_code=500, detail="Pinata API keys are missing on the backend.")

    try:
        # Read the file content
        file_content = await file.read()
        
        # Prepare the form data for Pinata
        # We use a tuple to specify (filename, file_content, content_type)
        files = {
            "file": (file.filename, file_content, file.content_type)
        }
        
        data = {}
        if pinataMetadata:
            data["pinataMetadata"] = pinataMetadata
        if pinataOptions:
            data["pinataOptions"] = pinataOptions

        headers = {
            "pinata_api_key": PINATA_API_KEY,
            "pinata_secret_api_key": PINATA_SECRET_KEY,
        }

        # Send to Pinata
        async with httpx.AsyncClient() as client:
            response = await client.post(
                "https://api.pinata.cloud/pinning/pinFileToIPFS",
                files=files,
                data=data,
                headers=headers,
                timeout=60.0 # Uploads can take a moment
            )

        if response.status_code != 200:
            error_msg = response.text
            print(f"Pinata Error: {response.status_code} - {error_msg}")
            raise HTTPException(status_code=502, detail=f"Pinata upstream error: {error_msg}")

        result = response.json()
        ipfs_hash = result.get("IpfsHash")
        
        if not ipfs_hash:
            raise HTTPException(status_code=502, detail="Pinata response did not contain an IpfsHash")

        return {
            "success": True,
            "cid": ipfs_hash,
            "url": f"{PINATA_GATEWAY}/{ipfs_hash}",
            "size": result.get("PinSize", 0)
        }

    except HTTPException:
        raise
    except Exception as e:
        print(f"Exception during IPFS upload: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to upload to IPFS: {str(e)}")
